alexandria/core/views.py

```diff
@@ -168,87 +171,63 @@ def multi(self, request, **kwargs):
 
             return response
 
-    @action(methods=["post"], detail=True, url_path="generate-thumbnail")
-    def generate_thumbnail(self, request, pk=None):
-        if not settings.ALEXANDRIA_ENABLE_THUMBNAIL_GENERATION:
-            raise ValidationError(_("Thumbnail generation is not enabled."))
-
-        file = self.get_object()
-        if file.variant == models.File.THUMBNAIL:
-            raise ValidationError(
-                _("File is already a thumbnail, cannot generate thumbnail.")
-            )
-        elif file.renderings.all().exists():
-            raise ValidationError(
-                _("File already has thumbnail, cannot generate multiple thumbnails.")
+    def create(self, request, *args, **kwargs):
+        serializer = self.get_serializer(data=request.data)
+        serializer.is_valid(raise_exception=True)
+        if settings.ALEXANDRIA_ENABLE_AT_REST_ENCRYPTION:
+            serializer.validated_data[
+                "encryption_status"
+            ] = settings.ALEXANDRIA_ENCRYPTION_METHOD
+        obj = serializer.save()
+        if (
+            settings.ALEXANDRIA_ENABLE_THUMBNAIL_GENERATION
+            and obj.variant == models.File.Variant.ORIGINAL
+        ):
+            try:
+                create_thumbnail(obj)
+            except DjangoCoreValidationError as e:
+                raise ValidationError(*e.messages)
+        if obj.variant == models.File.Variant.THUMBNAIL:
+            try:
+                # coerce the uploaded file to a thumbnail
+                create_thumbnail(obj)
+            except DjangoCoreValidationError as e:
+                # remove the uploaded file on failure to avoid
+                # collsions
+                obj.delete()
+                raise ValidationError(*e.messages)
+            # if the original already had a thumbnail remove that
+            self.queryset.filter(original=obj.original).exclude(pk=obj.pk).delete()
+        headers = self.get_success_headers(serializer.data)
+
+        return Response(serializer.data, status=HTTP_201_CREATED, headers=headers)
+
+    @permission_classes([AllowAny])
+    @action(methods=["get"], detail=True)
+    def download(self, request, pk=None):
+        if token_sig := request.query_params.get("signature"):
+            now = timezone.now()
+            expires = float(request.query_params.get("expires"))
```

Comment:
I would make this an `int`, not a `float` (also on URL generation side - no need to be this precise, and it can be a source of errors)

Reply:
done. it's all `int`s now